File: src/main/java/no/novari/discovery/service/mapping/InstanceMetadataContentMappingService.java
package no.novari.discovery.service.mapping;

import no.novari.discovery.service.model.dtos.InstanceMetadataContentDto;
import no.novari.discovery.service.model.entities.InstanceMetadataContent;
import org.springframework.stereotype.Service;

@Service
public class InstanceMetadataContentMappingService {

    private final InstanceValueMetadataMappingService instanceValueMetadataMappingService;
    private final InstanceObjectCollectionMetadataMappingService instanceObjectCollectionMetadataMappingService;
    private final InstanceMetadataCategoryMappingService instanceMetadataCategoryMappingService;

    public InstanceMetadataContentMappingService(
            InstanceValueMetadataMappingService instanceValueMetadataMappingService,
            InstanceObjectCollectionMetadataMappingService instanceObjectCollectionMetadataMappingService,
            InstanceMetadataCategoryMappingService instanceMetadataCategoryMappingService
    ) {
        this.instanceValueMetadataMappingService = instanceValueMetadataMappingService;
        this.instanceObjectCollectionMetadataMappingService = instanceObjectCollectionMetadataMappingService;
        this.instanceMetadataCategoryMappingService = instanceMetadataCategoryMappingService;
    }

    public InstanceMetadataContent toEntity(InstanceMetadataContentDto instanceMetadataContentDto) {
        return InstanceMetadataContent
                .builder()
                .instanceValueMetadata(
                        instanceMetadataContentDto.getInstanceValueMetadata()
                                .stream()
                                .map(instanceValueMetadataMappingService::toEntity)
                                .toList()
                )
                .instanceObjectCollectionMetadata(
                        instanceMetadataContentDto.getInstanceObjectCollectionMetadata()
                                .stream()
                                .map(instanceObjectCollectionMetadataMappingService::toEntity)
                                .toList()
                )
                .categories(
                        instanceMetadataContentDto.getCategories()
                                .stream()
                                .map(instanceMetadataCategoryMappingService::toEntity)
                                .toList()
                )
                .build();
    }

    public InstanceMetadataContentDto toDto(InstanceMetadataContent instanceMetadataContent) {
        return InstanceMetadataContentDto
                .builder()
                .instanceValueMetadata(
                        instanceMetadataContent.getInstanceValueMetadata()
                                .stream()
                                .map(instanceValueMetadataMappingService::toDto)
                                .toList()
                )
                .instanceObjectCollectionMetadata(
                        instanceMetadataContent.getInstanceObjectCollectionMetadata()
                                .stream()
                                .map(instanceObjectCollectionMetadataMappingService::toDto)
                                .toList()
                )
                .categories(
                        instanceMetadataContent.getCategories()
                                .stream()
                                .map(instanceMetadataCategoryMappingService::toDto)
                                .toList()
                )
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/discovery/service/mapping/InstanceMetadataContentMappingService.kt
package no.novari.flyt.discovery.service.mapping

import no.novari.flyt.discovery.service.model.dtos.InstanceMetadataContentDto
import no.novari.flyt.discovery.service.model.entities.InstanceMetadataContent
import org.springframework.stereotype.Service

@Service
class InstanceMetadataContentMappingService(
    private val instanceValueMetadataMappingService: InstanceValueMetadataMappingService,
    private val instanceObjectCollectionMetadataMappingService: InstanceObjectCollectionMetadataMappingService,
    private val instanceMetadataCategoryMappingService: InstanceMetadataCategoryMappingService,
) {
    fun toEntity(instanceMetadataContentDto: InstanceMetadataContentDto): InstanceMetadataContent {
        return InstanceMetadataContent(
            instanceValueMetadata =
                instanceMetadataContentDto.instanceValueMetadata
                    .map(instanceValueMetadataMappingService::toEntity)
                    .toMutableList(),
            instanceObjectCollectionMetadata =
                instanceMetadataContentDto.instanceObjectCollectionMetadata.map(
                    instanceObjectCollectionMetadataMappingService::toEntity,
                ).toMutableList(),
            categories =
                instanceMetadataContentDto.categories
                    .map(instanceMetadataCategoryMappingService::toEntity)
                    .toMutableList(),
        )
    }

    fun toDto(instanceMetadataContent: InstanceMetadataContent): InstanceMetadataContentDto {
        return InstanceMetadataContentDto(
            instanceValueMetadata =
                instanceMetadataContent.instanceValueMetadata.map(instanceValueMetadataMappingService::toDto),
            instanceObjectCollectionMetadata =
                instanceMetadataContent.instanceObjectCollectionMetadata.map(
                    instanceObjectCollectionMetadataMappingService::toDto,
                ),
            categories = instanceMetadataContent.categories.map(instanceMetadataCategoryMappingService::toDto),
        )
    }
}
